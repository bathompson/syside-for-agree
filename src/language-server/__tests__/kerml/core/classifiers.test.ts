/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

import { defaultLinkingErrorTo } from "../../../../testing";

describe.each(["specializes", "conjugates"])(
    "classifiers can only specialize other classifiers with '%s'",
    (token: string) => {
        test("specializing non-classifiers issues a diagnostic", async () => {
            return expect(
                `namespace A;
        classifier Child ${token} A;`
            ).toParseKerML(
                {},
                {
                    diagnostics: [defaultLinkingErrorTo("A")],
                }
            );
        });

        test("specializing classifiers is successful", async () => {
            return expect(
                `classifier A;
        classifier Child ${token} A;`
            ).toParseKerML({});
        });
    }
);

test("specializations can be parsed", async () => {
    return expect(`
    classifier A;
    classifier B;
    specialization Super subclassifier A specializes B;
    specialization subclassifier B :> A {
        doc /* unnamed */
    }`).toParseKerML("snapshot");
});

test("classifiers can specialize multiple other classifiers", async () => {
    return expect(`
    classifier A;
    classifier B;
    classifier C specializes A, B;`).toParseKerML("snapshot");
});
