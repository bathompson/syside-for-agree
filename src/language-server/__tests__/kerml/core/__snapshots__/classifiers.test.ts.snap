// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classifiers can specialize multiple other classifiers: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Classifier",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "A",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Classifier",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "B",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Classifier",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "C",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [
            {
              "$type": "Subclassification",
              "annotations": [],
              "chains": [],
              "elements": [],
              "prefixes": [],
              "reference": {
                "$type": "ClassifierReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Classifier",
                      "aliases": [],
                      "annotatingMembers": [],
                      "declaredName": "A",
                      "filters": [],
                      "imports": [],
                      "isSufficient": false,
                      "members": [],
                      "namespaceMembers": [],
                      "prefixes": [],
                      "relationshipMembers": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
            {
              "$type": "Subclassification",
              "annotations": [],
              "chains": [],
              "elements": [],
              "prefixes": [],
              "reference": {
                "$type": "ClassifierReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Classifier",
                      "aliases": [],
                      "annotatingMembers": [],
                      "declaredName": "B",
                      "filters": [],
                      "imports": [],
                      "isSufficient": false,
                      "members": [],
                      "namespaceMembers": [],
                      "prefixes": [],
                      "relationshipMembers": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
          ],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;

exports[`specializations can be parsed: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Classifier",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "A",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Classifier",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "B",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Subclassification",
          "annotations": [],
          "chains": [],
          "declaredName": "Super",
          "elements": [],
          "prefixes": [],
          "reference": {
            "$type": "ClassifierReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Classifier",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "B",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "ClassifierReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Classifier",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "A",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Subclassification",
          "annotations": [
            {
              "$type": "Annotation",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Documentation",
                "about": [],
                "body": "/* unnamed */",
                "prefixes": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "chains": [],
          "elements": [],
          "prefixes": [],
          "reference": {
            "$type": "ClassifierReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Classifier",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "A",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "ClassifierReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Classifier",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "B",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
        "prefixes": [],
      },
    ],
  },
}
`;
